File: Computer Networks/Question 3/sorting.py
def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quick_sort(left) + middle + quick_sort(right)

def merge_sort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        left = arr[:mid]
        right = arr[mid:]

        merge_sort(left)
        merge_sort(right)

        i = j = k = 0
        while i < len(left) and j < len(right):
            if left[i] < right[j]:
                arr[k] = left[i]
                i += 1
            else:
                arr[k] = right[j]
                j += 1
            k += 1

        while i < len(left):
            arr[k] = left[i]
            i += 1
            k += 1

        while j < len(right):
            arr[k] = right[j]
            j += 1
            k += 1
    return arr

def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and key < arr[j]:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr

def read_file(filename):
    with open(filename, "r") as file:
        data = list(map(int, file.read().split()))
    return data

def append_to_file(filename, data):
    with open(filename, "a") as file:
        file.write("\n" + " ".join(map(str, data)))

def main():
    filename = "data.txt"
    data = read_file(filename)

    # Quick Sort
    sorted_data = quick_sort(data.copy())
    append_to_file(filename, sorted_data)

    # Merge Sort
    sorted_data = merge_sort(data.copy())
    append_to_file(filename, sorted_data)

    # Insertion Sort
    sorted_data = insertion_sort(data.copy())
    append_to_file(filename, sorted_data)

    print("Sorting completed and appended to the file.")

if __name__ == "__main__":
    main()

File: Computer Networks/Question 3/sorting.c
#include <stdio.h>
#include <stdlib.h>
void quickSort(int arr[], int low, int high);
int partition(int arr[], int low, int high);
void mergeSort(int arr[], int left, int right);
void merge(int arr[], int left, int mid, int right);
void insertionSort(int arr[], int n);
void readFromFile(const char *filename, int arr[], int *size);
void appendToFile(const char *filename, int arr[], int size);
int main() {
    const char *filename = "data.txt";
    int arr[1000], size;
    readFromFile(filename, arr, &size);
    quickSort(arr, 0, size - 1);
    appendToFile(filename, arr, size);
    readFromFile(filename, arr, &size);
    mergeSort(arr, 0, size - 1);
    appendToFile(filename, arr, size);
    readFromFile(filename, arr, &size);
    insertionSort(arr, size);
    appendToFile(filename, arr, size);

    printf("Sorting completed and appended to the file.\n");
    return 0;
}

void quickSort(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = (low - 1);
    for (int j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
    int temp = arr[i + 1];
    arr[i + 1] = arr[high];
    arr[high] = temp;
    return i + 1;
}

void mergeSort(int arr[], int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

void merge(int arr[], int left, int mid, int right) {
    int n1 = mid - left + 1;
    int n2 = right - mid;
    int L[n1], R[n2];

    for (int i = 0; i < n1; i++)
        L[i] = arr[left + i];
    for (int j = 0; j < n2; j++)
        R[j] = arr[mid + 1 + j];

    int i = 0, j = 0, k = left;
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }

    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }
}

void insertionSort(int arr[], int n) {
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

void readFromFile(const char *filename, int arr[], int *size) {
    FILE *file = fopen(filename, "r");
    if (!file) {
        printf("Error opening file!\n");
        exit(1);
    }
    *size = 0;
    while (fscanf(file, "%d", &arr[*size]) != EOF) {
        (*size)++;
    }
    fclose(file);
}

void appendToFile(const char *filename, int arr[], int size) {
    FILE *file = fopen(filename, "a");
    if (!file) {
        printf("Error opening file!\n");
        exit(1);
    }
    fprintf(file, "\n");
    for (int i = 0; i < size; i++) {
        fprintf(file, "%d ", arr[i]);
    }
    fclose(file);
}